server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollector.java

```diff
@@ -177,11 +183,39 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
                     throw e;
                 }
             }
-            aggsLeafCollector = null;
         }
-        // say that the aggs collector early terminates while the top docs collector does not, we still want to wrap in the same way
+        final LeafCollector aggsLeafCollector = alf;
+
+        if (topDocsLeafCollector == null && terminateAfter == 0 && minScore == null) {
+            // top docs collector early terminated, we can avoid wrapping as long as we don't need to apply terminate_after and min_score
+            // aggs don't support skipping low scoring hits, so we can rely on setMinCompetitiveScore being a no-op already
+            return aggsLeafCollector;
+        }
```

Comment:
Similar here, we should check `postFilterBits` is null or at a minimum assert such.

Reply:
we don't check for post filter bits here, nor post filter weight, because it does not matter. Post filter is applied to top docs only and top docs collection has early terminated. I can add a comment around this.